ggml/src/gguf.cpp

```diff
@@ -0,0 +1,1311 @@
+            gguf_free(ctx);
+            return nullptr;
+        }
+
+        struct ggml_context * ctx_data = *params.ctx;
+
+        struct ggml_tensor * data = nullptr;
+
+        if (!params.no_alloc) {
+            data = ggml_new_tensor_1d(ctx_data, GGML_TYPE_I8, ctx->size);
+
+            ok = ok && data != nullptr;
+
+            // read the binary blob with the tensor data
+            ok = ok && gr.read(data->data, ctx->size);
+
+            if (!ok) {
+                fprintf(stderr, "%s: failed to read tensor data binary blob\n", __func__);
+                ggml_free(ctx_data);
+                *params.ctx = nullptr;
+                gguf_free(ctx);
+                return nullptr;
+            }
+
+            ctx->data = data->data;
+        }
+
+        ggml_set_no_alloc(ctx_data, true);
+
+        // create the tensors
+        for (uint64_t i = 0; i < n_tensors; ++i) {
+            struct ggml_tensor * cur = ggml_new_tensor(
+                ctx_data, ctx->info[i].t.type, GGML_MAX_DIMS, ctx->info[i].t.ne);
+
+            ok = ok && cur != nullptr;
+
+            if (!ok) {
+                break;
+            }
+
+            ggml_set_name(cur, ctx->info[i].t.name);
+
+            // point the data member to the appropriate location in the binary blob using the tensor info
+            if (!params.no_alloc) {
+                cur->data = (char *) data->data + ctx->info[i].offset;
+            }
+        }
+
+        if (!ok) {
+            fprintf(stderr, "%s: failed to create tensors\n", __func__);
+            ggml_free(ctx_data);
+            *params.ctx = nullptr;
+            gguf_free(ctx);
+            return nullptr;
+        }
+
+        ggml_set_no_alloc(ctx_data, params.no_alloc);
+    }
+
+    return ctx;
+}
+
+struct gguf_context * gguf_init_from_file(const char * fname, struct gguf_init_params params) {
+    std::ifstream ifs(fname, std::ios::binary);
+    return gguf_init_from_file_impl(ifs, params);
+}
+
+void gguf_free(struct gguf_context * ctx) {
+    if (ctx == nullptr) {
+        return;
+    }
+    delete ctx;
+}
+
+const char * gguf_type_name(enum gguf_type type) {
+    auto it = GGUF_TYPE_NAME.find(type);
+    return it == GGUF_TYPE_NAME.end() ? nullptr : it->second;
+}
+
+int gguf_get_version(const struct gguf_context * ctx) {
+    return ctx->version;
+}
+
+size_t gguf_get_alignment(const struct gguf_context * ctx) {
+    return ctx->alignment;
+}
+
+size_t gguf_get_data_offset(const struct gguf_context * ctx) {
+    return ctx->offset;
+}
+
+// TODO should this be a const pointer? should it exist at all?
+void * gguf_get_data(const struct gguf_context * ctx) {
+    return ctx->data;
+}
+
+// TODO this returns int but the underlying type is uint64
+int gguf_get_n_kv(const struct gguf_context * ctx) {
+    return ctx->kv.size();
+}
+
+// TODO should this return uint64_t?
+int gguf_find_key(const struct gguf_context * ctx, const char * key) {
+    // return -1 if key not found
+    int keyfound = -1;
```

Comment:
Yes, I think so. In general, we use `size_t` for memory sizes and offsets expressed in bytes and `int64_t` for counting things like number of elements. The GGUF spec slightly deviates from the convention and uses `uint64_t` for both of these, which is not ideal. But at least the interface should match the underlying types.

Also, there are GGUF functions that use `int` for indexing KV and tensors that would also need to be updated:

```c++
    GGML_API uint8_t      gguf_get_val_u8  (const struct gguf_context * ctx, int key_id);

    GGML_API const char *   gguf_get_tensor_name  (const struct gguf_context * ctx, int i);
```

However, this might become a breaking change. Maybe we should do this in a follow-up PR?

Reply:
Changing the data type from a signed to an unsigned integer would definitely be a breaking change since the intended way to check whether a key exists is to check whether the returned index is non-negative. It would also be possible to return `int64_t` since the maximum number of `gguf_kv`/`gguf_tensor_info` elements that can be allocated on a 64 bit system is going to be smaller than the maximum value that can be represented with `int64_t`.